android/app/src/main/kotlin/com/zulip/flutter/ZulipPlugin.kt

```diff
@@ -69,6 +85,87 @@ private class AndroidNotificationHost(val context: Context)
         NotificationManagerCompat.from(context).deleteNotificationChannel(channelId)
     }
 
+    override fun listStoredSoundsInNotificationsDirectory(): List<StoredNotificationsSound> {
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
+            throw UnsupportedOperationException()
+        }
+
+        // Query and cursor-loop based on:
+        //   https://developer.android.com/training/data-storage/shared/media#query-collection
+        val collection = AudioStore.getContentUri(MediaStore.VOLUME_EXTERNAL_PRIMARY)
+        val projection = arrayOf(AudioStore._ID, AudioStore.DISPLAY_NAME, AudioStore.OWNER_PACKAGE_NAME)
+        val selection = "${AudioStore.RELATIVE_PATH}=?"
+        val selectionArgs = arrayOf(notificationSoundsDirectoryPath)
+        val sortOrder = "${AudioStore._ID} ASC"
+
+        val sounds = mutableListOf<StoredNotificationsSound>()
+        val query = context.contentResolver.query(
+            collection,
+            projection,
+            selection,
+            selectionArgs,
+            sortOrder,
+        )
+        query?.use { cursor ->
```

Comment:
> inserting duplicates, which could then pile up.

(This shouldn't happen, because once we successfully create the channel we won't be entering this whole sound-file-management area of code on future launches… but once things are going wrong in our interaction with the platform, maybe there's also a failure that we hit before the point of successfully creating the channel.)

Reply:
> or perhaps throw an exception. Then the caller on the Dart side can see that and abort with `return defaultSoundUrl`.

Updated with this.